File: Python/all_in_one.py
import datetime
import random
import string 
import sys

class LOG :
    CEND      = '\33[0m'
    CBOLD     = '\33[1m'
    CITALIC   = '\33[3m'
    CURL      = '\33[4m'
    CBLINK    = '\33[5m'
    CBLINK2   = '\33[6m'
    CSELECTED = '\33[7m'

    CBLACK  = '\33[30m'
    CRED    = '\33[31m'
    CGREEN  = '\33[32m'
    CYELLOW = '\33[33m'
    CBLUE   = '\33[34m'
    CVIOLET = '\33[35m'
    CBEIGE  = '\33[36m'
    CWHITE  = '\33[37m'

    CBLACKBG  = '\33[40m'
    CREDBG    = '\33[41m'
    CGREENBG  = '\33[42m'
    CYELLOWBG = '\33[43m'
    CBLUEBG   = '\33[44m'
    CVIOLETBG = '\33[45m'
    CBEIGEBG  = '\33[46m'
    CWHITEBG  = '\33[47m'

    CGREY    = '\33[90m'
    CRED2    = '\33[91m'
    CGREEN2  = '\33[92m'
    CYELLOW2 = '\33[93m'
    CBLUE2   = '\33[94m'
    CVIOLET2 = '\33[95m'
    CBEIGE2  = '\33[96m'
    CWHITE2  = '\33[97m'

    CGREYBG    = '\33[100m'
    CREDBG2    = '\33[101m'
    CGREENBG2  = '\33[102m'
    CYELLOWBG2 = '\33[103m'
    CBLUEBG2   = '\33[104m'
    CVIOLETBG2 = '\33[105m'
    CBEIGEBG2  = '\33[106m'
    CWHITEBG2  = '\33[107m'


def fill_availableRows(board_size):
    availableRows = []
    for x in range(board_size):
        availableRows.append(x)
    return availableRows

def print_solution(queens):

    print(f'{LOG.CVIOLET}---- This is your solution ----{LOG.CEND}')
    alphabet = []

    for letter in range(len(queens) if len(queens)<=26 else 26):
        alphabet.append(f' {string.ascii_uppercase[letter]}')
    if(len(queens) > 26):
        for letter in range(len(queens)-26):
            alphabet.append(f' {string.ascii_lowercase[letter]}')

    isColored = True
    print('  ', *alphabet)

    for i in range(len(queens)):
        row = []
        if len(queens) % 2 == 0:
            isColored = not(isColored)

        for j in range(len(queens)):
            isColored = not(isColored)
            if queens[j] == i:
                row.append(f'{LOG.CREDBG}{LOG.CBLACK if isColored else LOG.CWHITE} 1') 
            else:
                row.append(f'{LOG.CWHITEBG if isColored else LOG.CBLACKBG}{LOG.CBLACK if isColored else LOG.CWHITE} 0')

        print(f'{LOG.CEND}{str(i+1).zfill(2)}', *row, LOG.CEND)


def validator(queens, positive_diagonal, negative_diagonal, available_rows, row, column, board_size):
    if not(available_rows[row] in queens) and not(((board_size-1) - available_rows[row] - column) in positive_diagonal ) and not( (column - available_rows[row]) in negative_diagonal):
        return True
    return False

def placing_queens(queens, positive_diagonal, negative_diagonal, available_rows, column, board_size ):
    laps = 0
    row = random.randint(0, len(available_rows))

    for row in range(len(available_rows)) :
        if validator(queens, positive_diagonal, negative_diagonal, available_rows, row, column, board_size) :

            queens.append(available_rows[row])
            positive_diagonal.append((board_size-1)-available_rows[row]-column)
            negative_diagonal.append(column-available_rows[row])
            value = available_rows.pop(row)

            if len(queens) == board_size: return [True, queens]

            [status, returnedQueens] = placing_queens(queens, positive_diagonal, negative_diagonal, available_rows, column+1, board_size)

            if status : return [status, returnedQueens]

            available_rows.insert(row, value)
            queens.pop()
            positive_diagonal.pop()
            negative_diagonal.pop()

        if row + 1 == len(available_rows) and laps < 1: 
            row = -1 
            laps += 1

    return [False, queens]


def n_queens(board_size):
    if board_size<4 and board_size != 1 : 
        return "Not posible!"

    available_rows = fill_availableRows(board_size)
    column = 0
    negative_diagonal = []
    positive_diagonal = []
    queens = []

    [status, returnedQueens] = placing_queens( queens, positive_diagonal, negative_diagonal, available_rows, column, board_size)

    if not(status) :
        return 'It was not posible' 

    return returnedQueens


for argument in range(1, len(sys.argv)):
    print(f'\n{LOG.CBLUE}****************N-Queens****************{LOG.CEND}')
    print(f'Size of Board: {LOG.CWHITE}{sys.argv[argument]}{LOG.CEND}')

    startTime = datetime.datetime.now()
    queens = n_queens(int(sys.argv[argument]))
    elapsedSeconds = (datetime.datetime.now() - startTime)
    
    if type(queens) != type('string') :
        print(f'Seconds Elapsed: {LOG.CGREEN}{elapsedSeconds}{LOG.CEND}')
        print_solution(queens)

        print("\nQueens positions by column:")
        print(queens) 

    if type(queens) == type('string'):
        print(LOG.CRED, queens, LOG.CEND)

File: JavaScript/allInOne.js
// Constant for Log Style
const LOG = {
    reset: "\x1b[0m",
    bright: "\x1b[1m",
    dim: "\x1b[2m",
    underscore: "\x1b[4m",
    blink: "\x1b[5m",
    reverse: "\x1b[7m",
    hidden: "\x1b[8m",
    // Foreground (text) colors
    fg: {
        black: "\x1b[30m",
        red: "\x1b[31m",
        green: "\x1b[32m",
        yellow: "\x1b[33m",
        blue: "\x1b[34m",
        magenta: "\x1b[35m",
        cyan: "\x1b[36m",
        white: "\x1b[37m",
        crimson: "\x1b[38m"
    },
    // Background colors
    bg: {
        black: "\x1b[40m",
        red: "\x1b[41m",
        green: "\x1b[42m",
        yellow: "\x1b[43m",
        blue: "\x1b[44m",
        magenta: "\x1b[45m",
        cyan: "\x1b[46m",
        white: "\x1b[47m",
        crimson: "\x1b[48m"
    }
};

/**
 * fillAvailableRows
 * Complexity: O(n)
 * @param {integer} boardSize 
 * @returns {Array}
 */
function fillAvailableRows(boardSize){
    let availableRows = new Array();
    for(let index = 0; index<boardSize; index++){
        availableRows.push(index);
    }
    return availableRows
}

/**
 * Validation
 * Prints the board with the queens as 1, and 0 for an empty space
 * Complexity O(n^2)
 * @param {Set} queens 
 * @returns {void}
 */
function printSolution(queens){
    console.log(`${LOG.fg.magenta}---- This is your solution ----${LOG.reset}`);
    const alphabetASCII = Array.from(Array(queens.size)).map((e, i) => i + 65);
    const alphabet = alphabetASCII.map((x) => ` ${String.fromCharCode(x)}`);
    let isColored = true;
    console.log(`  `, ...alphabet);

    for(let i=0; i<queens.size; i++){
        let iterator = queens.values();
        let row = new Array();
        queens.size % 2 == 0 ? isColored = !isColored : {}
        for(let j = 0; j<queens.size; j++){
            isColored = !isColored
            iterator.next().value == i ? 
                row.push(`${LOG.bg.red}${isColored ? LOG.fg.black : LOG.fg.white} 1`) 
                : row.push(`${isColored? LOG.bg.white : LOG.bg.black}${isColored? LOG.fg.black : LOG.fg.white} 0`)
        }
        console.log(`${(i+1).toString().padStart(2, 0)}`, ...row, `${LOG.reset}`);
    }
}

const validate = (state, row, column, boardSize) => {
    return  ( !state.queens.has(state.availableRows[row]) &&
            !state.positiveDiagonal.has((boardSize-1) - state.availableRows[row] - column) &&
            !state.negativeDiagonal.has(column - state.availableRows[row]) )
}

/**
 * 
 * @param {Object} state 
 * @param {integer} column 
 * @param {integer} boardSize 
 * @returns 
 */
function PlacingQueens( state, column, boardSize ){
    let laps = 0;
    let row = Math.floor(Math.random() * (state.availableRows.length - 0));

    for(row; row < state.availableRows.length; row++){

        if( validate(state, row, column, boardSize) ){

            state.queens.add(state.availableRows[row]);
            state.positiveDiagonal.add((boardSize-1)-state.availableRows[row]-column);
            state.negativeDiagonal.add(column-state.availableRows[row]);
            let [value] = state.availableRows.splice(row,1);

            if(state.queens.size == boardSize) return [true, state.queens]

            var [status, returnedQueens] = PlacingQueens(state, column+1, boardSize)

            if(status) return [status, returnedQueens]

            state.availableRows.splice(row,0, value)
            state.queens.delete(state.availableRows[row]);
            state.positiveDiagonal.delete((boardSize-1)-state.availableRows[row]-column);
            state.negativeDiagonal.delete(column-state.availableRows[row]);
        }

        if(row + 1 == state.availableRows.length && laps < 1) row = -1, laps += 1;
    }

    return [false, state.queens]
}

/**
 * Queens
 * Starting Function, used for creating variables for the placing queens function
 * @param {int} boardSize 
 * @returns 
 */
function Queens(boardSize){

    if(boardSize<4 && boardSize != 1) return `Not posible! ${boardSize} < 4 and it's different of 1`

    let availableRows = fillAvailableRows(boardSize);
    let column = 0
    var negativeDiagonal = new Set(); // [ -|n-1|, n-1 ]
    var positiveDiagonal = new Set(); // [ -|n-1|, n-1 ]
    var queens = new Set(); // Max length = n

    const state = {
        availableRows,
        negativeDiagonal,
        positiveDiagonal,
        queens
    }

    let [status, returnedQueens] = PlacingQueens(state, column, boardSize)

    return !status ? 'It was not posible' : returnedQueens;
}

var myArgs = process.argv.slice(2);

function main(myArgs){
    myArgs.map( e => {

        console.log(`\n${LOG.fg.cyan}****************N-Queens****************${LOG.reset}`);
        console.log(`Size of Board: ${LOG.fg.white}${e}${LOG.reset}`);

        const startTime = Date.now();
        let queens = Queens(e)
        const elapsedSeconds = (Date.now() - startTime)/1000;

        if(typeof queens !== 'string'){
            console.log(`Seconds Elapsed: ${LOG.fg.green}${elapsedSeconds}${LOG.reset}`);
            
            printSolution(queens);
            
            console.log(`\nQueens positions by column (0-${e-1}):`);
            console.log(queens); 
        }

        typeof queens === 'string' ? console.log(LOG.fg.red, queens, LOG.reset) : {}
    })
}
